### ISSUE TITLE
Array errors seemingly related to presence of outliers

### ISSUE THREAD

**Issue by dwysocki** [astroswego/plotypus#94]
There are two errors I have encountered, which I believe are related to the presence of outliers.

When there _are_ outliers, the following error seems to happen regardless of output types.

```
$ plotypus -i data/ogle-iii/I/OGLE-LMC-CEP-0002.dat --period 3.2 --sigma 5
Name    A_0 dA_0    Period  Shift   Coverage    Inliers Outliers    R^2 MSE MaxDegree   Params
Traceback (most recent call last):
  File "/usr/bin/plotypus", line 9, in <module>
    load_entry_point('plotypus==0.2.6.dev0', 'console_scripts', 'plotypus')()
  File "/home/dwysocki/research/vstars/plotypus/src/plotypus/cli.py", line 414, in main
    processes=args.star_processes, **args_dict)
  File "/home/dwysocki/research/vstars/plotypus/src/plotypus/utils.py", line 103, in pmap
    result = func(arg, **kwargs)
  File "/home/dwysocki/research/vstars/plotypus/src/plotypus/cli.py", line 593, in process_star
    **kwargs)
  File "/home/dwysocki/research/vstars/plotypus/src/plotypus/lightcurve.py", line 429, in get_lightcurve_from_file
    verbosity=verbosity, **kwargs)
  File "/home/dwysocki/research/vstars/plotypus/src/plotypus/lightcurve.py", line 313, in get_lightcurve
    residuals = np.column_stack((data.T[0], phase, data.T[1], residuals, data.T[2]))
  File "/usr/lib/python3.5/site-packages/numpy/lib/shape_base.py", line 317, in column_stack
    return _nx.concatenate(arrays, 1)
ValueError: all the input array dimensions except for the concatenation axis must match exactly
```

If I then increase the number of sigmas required to be flagged as an outlier, such that none are detected, the error vanishes.

```
$ plotypus -i data/ogle-iii/I/OGLE-LMC-CEP-0002.dat --period 3.2 --sigma 100
Name    A_0 dA_0    Period  Shift   Coverage    Inliers Outliers    R^2 MSE MaxDegree   Params
OGLE-LMC-CEP-0002   15.6661 0.00135085  3.2 0.1 0.96    366 0   -0.0229705  0.00730041  15  9
```

However, there is still another error in this case, which I believe to be a result of manipulating a size 0 array when trying to plot outliers.

```
$ plotypus -i data/ogle-iii/I/OGLE-LMC-CEP-0002.dat --period 3.2 --sigma 100 -o out
Name    A_0 dA_0    Period  Shift   Coverage    Inliers Outliers    R^2 MSE MaxDegree   Params
Traceback (most recent call last):
  File "/usr/bin/plotypus", line 9, in <module>
    load_entry_point('plotypus==0.2.6.dev0', 'console_scripts', 'plotypus')()
  File "/home/dwysocki/research/vstars/plotypus/src/plotypus/cli.py", line 414, in main
    processes=args.star_processes, **args_dict)
  File "/home/dwysocki/research/vstars/plotypus/src/plotypus/utils.py", line 103, in pmap
    result = func(arg, **kwargs)
  File "/home/dwysocki/research/vstars/plotypus/src/plotypus/cli.py", line 675, in process_star
    **ChainMap(kwargs, result))
  File "/home/dwysocki/research/vstars/plotypus/src/plotypus/lightcurve.py", line 606, in plot_lightcurve
    return(plot_lightcurve_mpl(*args, **kwargs))
  File "/home/dwysocki/research/vstars/plotypus/src/plotypus/lightcurve.py", line 683, in plot_lightcurve_mpl
    capsize=0 if color else 1)
  File "/usr/lib/python3.5/site-packages/matplotlib/__init__.py", line 1811, in inner
    return func(ax, *args, **kwargs)
  File "/usr/lib/python3.5/site-packages/matplotlib/axes/_axes.py", line 2958, in errorbar
    iterable(yerr[0]) and len(yerr[0]) > 1))):
IndexError: index 0 is out of bounds for axis 0 with size 0
```

**Comment by dwysocki** [comment]
I've identified the source of the error when there are _no_ outliers. The problem is calling `pyplot.errorbar` and providing a length zero array to `yerr`. Here is a self-contained reproduction of the error, which also demonstrates that the error does not occur if `yerr` is not provided.

``` python
In [1]: import numpy as np

In [2]: import matplotlib.pyplot as plt

In [3]: fig, ax = plt.subplots()

In [4]: empty_array = np.array([])

In [5]: ax.errorbar(empty_array, empty_array) # no problem
Out[5]: <Container object of 3 artists>

In [6]: ax.errorbar(empty_array, empty_array, yerr=empty_array) # problem
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-6-3f3a6bcfee03> in <module>()
----> 1 ax.errorbar(empty_array, empty_array, yerr=empty_array) # problem

/usr/lib/python3.5/site-packages/matplotlib/__init__.py in inner(ax, *args, **kwargs)
   1809                     warnings.warn(msg % (label_namer, func.__name__),
   1810                                   RuntimeWarning, stacklevel=2)
-> 1811             return func(ax, *args, **kwargs)
   1812         pre_doc = inner.__doc__
   1813         if pre_doc is None:

/usr/lib/python3.5/site-packages/matplotlib/axes/_axes.py in errorbar(self, x, y, yerr, xerr, fmt, ecolor, elinewidth, capsize, barsabove, lolims, uplims, xlolims, xuplims, errorevery, capthick, **kwargs)
   2956                 if not (len(yerr) == 1 or
   2957                         (len(yerr) == len(y) and not (
-> 2958                             iterable(yerr[0]) and len(yerr[0]) > 1))):
   2959                     raise ValueError("yerr must be a scalar, the same "
   2960                                      "dimensions as y, or 2xN.")

IndexError: index 0 is out of bounds for axis 0 with size 0
```

As an aside: I also tried this with empty _lists_ instead of arrays, and I received a similar but different error. The real problem here is that `pyplot.errorbar` does not properly validate its inputs, and so the exception is raised when it tries an invalid operation on the object given to `yerr`.

**Comment by dwysocki** [comment]
I've now identified part of the issue when there _are_ outliers. I identified the line in the stacktrace where the exception was being thrown, and added a print statement, displaying the shapes of the arrays being referenced.

``` python
    # use rephased phase points from *data* in residuals
    print("data = {}, phase = {}, residuals = {}"
          .format(data.shape, phase.shape, residuals.shape),
          file=stderr)
    exit()
    residuals = np.column_stack((data.T[0], phase, data.T[1], residuals, data.T[2]))
```

The output is as follows (using the same calling sequence as before)

``` python
data = (366, 3), phase = (366,), residuals = (363,)
```

So I'm assuming that `residuals` only contains inliers, while `data` and `phase` simply _masked_ the outliers, and contains both. This probably has to do with the fact that `residuals` is being sampled from a predictor object in

``` python
    residuals = prediction_residuals(phase, mag, predictor)
```

so I need to look into how the `prediction_residuals` function is behaving. Presumably it only returns predictions for unmasked values. Ideally it should make predictions for masked _and_ unmasked values, and simply apply the same mask to the predictions.

**Comment by dwysocki** [comment]
Under further inspection, the problem is not with the `prediction_residuals` function, but its input. It is provided the `phase` and `mag` arrays, which were obtained from `utils.get_signal` (i.e. they contain no outliers). Instead we need to provide masked arrays, containing both inliers and outliers.

I believe this is a symptom of a bigger problem, however. I think `get_signal` is overused, and we should be preserving masked arrays much more often than we do. So for now I am going to fix the symptom, so that things are at least functioning. In the future, the problem will be fixed in #96.

**Comment by dwysocki** [comment]
Fixed in #97
